src/sqlfluff/core/parser/grammar/conditional.py

```diff
@@ -0,0 +1,67 @@
+"""Conditional Grammar."""
+
+from sqlfluff.core.parser.segments import Indent
+from sqlfluff.core.parser.match_result import MatchResult
+from sqlfluff.core.parser.match_wrapper import match_wrapper
+
+from sqlfluff.core.parser.grammar.base import (
+    BaseGrammar,
+)
+
+
+class Conditional(BaseGrammar):
+    """A grammar which is conditional on the parse context.
+
+    Args:
+        enforce_whitespace_preceeding (:obj:`bool`): Should the GreedyUntil
+            match only match the content if it's preceded by whitespace?
+            (defaults to False). This is useful for some keywords which may
+            have false alarms on some array accessors.
+
+    """
+
+    def __init__(self, *args, config_type: str = "", **kwargs):
+        if not all(issubclass(arg, Indent) for arg in args):
+            raise ValueError(
+                "Conditional is only designed to work with Indent segments."
+            )
+        if len(args) != 1:
+            raise ValueError(
+                "Conditional is only designed to work with a single element."
+            )
+        if not config_type:
+            raise ValueError("Conditional config_type must be set.")
+        if not kwargs:
+            raise ValueError("Conditional requires rules to be set.")
+        self._config_type = config_type
+        self._config_rules = kwargs
+        super().__init__(*args)
+
+    def is_enabled(self, parse_context):
+        """Evaluate conditionals and return whether enabled."""
+        # If no config rules are set then it's always enabled.
+        try:
+            config = {"indent": parse_context.indentation_config}[self._config_type]
+        except KeyError:

```

Comment:
This code reads a bit strangely to me. IIUC, it's doing two things:
* Verifying that `self._config_type` is `"indent"`
* Assigning `config = parse_context.indentation_config`

Could we do this separately? E.g.
```
if self._config_type != "indent":
    raise ValueError(...)
config = parse_context.indentation_config
```

I think this version is easier to read.

Reply:
you read it correctly. The idea here is that we might have other kinds of config other than "indent" might be possible in future. The code is kind of hard to read though, I'll split it up a bit.

Reply:
Ok, I've simplified it a load, and taken out some of the logic around other configuration paths for now. I've extended the documentation a lot to describe where we those points might be. Would love your views on the changes.

Reply:
Sure! I'll try and take a look today.